자동차 경주 구현 step1 제출합니다

BurningFalls opened the pull request:
페어와 너무 열정적으로 코딩하다보니, 기능 단위로 커밋하는 것을 중도에 알아차려서 처음 커밋들이 너무 큰 단위로 묶여있는점 양해부탁드립니다.

기본적으로 MVC 패턴을 지키려고 노력했고, 자바 컨벤션을 최대한 사용해봤습니다.

`Cars` 클래스에 `List<Car>`가 들어있어서, 여러 자동차에 대한 처리를 해야하면 모두 `Cars` 클래스에서 하도록 만들었습니다. 

리뷰 부탁드립니다. 감사합니다.

BurningFalls committed:
docs: 구현할 기능 목록 작성

BurningFalls committed:
feat+test: trycount 기능 구현 및 테스트 완료

BurningFalls committed:
feat+test: car 기능 구현 및 테스트 완료

BurningFalls committed:
feat+test: cars 기능 구현 및 테스트 완료

BurningFalls committed:
feat: InputView, OutputView 구현 완료

BurningFalls committed:
fix: InputView 예외 직접 처리

BurningFalls committed:
style: 함수 이름 변경, 개행문자 추가

BurningFalls committed:
feat: 자동차 게임 구현 완료

BurningFalls committed:
style: 미사용 import문 삭제 및 자바 컨벤션 적용

BurningFalls committed:
style: 오타 수정

BurningFalls committed:
style: 기능 테스트 이름 변경 (Function -> Feature)

BurningFalls committed:
style: @DisplayName 위치 변경

pci2676 commented on src/main/java/racingcar/controller/CarController.java:
```diff
@@ -0,0 +1,29 @@
+import racingcar.view.OutputView;
+
+public class CarController {
+    private Cars cars;
+    private TryCount tryCount;
```
`CarController`의 멤버변수로 관리해야하는 이유가 있었나요?

BurningFalls replied:
`startGame` 메서드 내부에서만 해당 변수들을 사용했기 때문에, 해당 인스턴스 변수들을 메서드 내부의 지역 변수로 관리하는 것이 더 좋네요! 미처 생각하지 못했던 부분이었습니다.

pci2676 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private static void nameUnderFive(String name) {
+        if (name.length() > 5) {
```
매직 넘버를 다르게 표현해보면 좋을 것 같습니다.

https://javabom.tistory.com/28

BurningFalls replied:
매직넘버를 상수로 선언해서 가독성과 유연성을 높이는 자바 컨벤션을 지키지 않았었네요. 감사합니다!

pci2676 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%s : %s", name, "-".repeat(position));
+    }
```
toString을 이용해서 view에 대한 표현을 하신것 같습니다.

domain이 view를 의존하게 되는것으로 보이는데 의도하신 방향일까요?

BurningFalls replied:
1. `OutputView`에서 단순히 `sout(car)`만 작성해도 되니까 코드가 깔끔해서 좋아.
2. `toString`을 사용하지 않으면 `OutputView`에서 인수로 받은 객체의 멤버 변수에 접근하기 위해서 `get`을 많이 사용해야하는데, `get`의 사용을 최소화하고 싶어.

해당 코드를 작성할 당시, 위의 두 마음가짐을 갖고 있었고, domain이 view를 의존하게 되어 MVC 패턴의 원칙을 어기는 것에 대해서는 생각해보지 못했습니다. 아래는 제가 조사해본 내용입니다.

* `toString()` 구현은 주로 간단한 문자열 표현을 제공하는 데 목적이 있다. 이는 뷰가 사용자에게 데이터를 어떻게 표시할지 결정하는 것과는 별개이다.
* `toString()` 메서드는 모델의 문자열 표현을 제공하는 일반적인 방법일 뿐, `UI`를 구성하는 데 사용되어서는 안 된다. 따라서, 뷰는 모델의 데이터를 직접적으로 처리하여 사용자에게 적절한 형태로 표시하는 역할을 담당한다.
* `toString()` 메서드는 모델이 뷰를 의존하게 만드는 것이 아니라, 객체의 상태를 문자열로 간단히 표현하는 유틸리티로 간주된다. `toString()`의 사용은 이러한 원칙에 영향을 주지 않으며, 올바르게 사용된다면 MVC 패턴 분리 원칙을 해치지 않는다.

조사한 내용을 바탕으로 제가 이해한 내용은 아래와 같습니다.

* `toString()`에 많은 역할을 위임하여 `OutputView`의 코드를 단순하게 만드는 것이 좋다라고 생각하는 것은 옳지 않다. 사용자에게 어떻게 표현할지 결정하는 것은 `View`가 해야 할 역할과 책임이다.
* `toString()`을 위 코드와 같이 깊게 정의해버리면, 사용자가 `Car`에 대해서 다른 방식으로 표현하고 싶어졌을 때 또다른 `toString()` 함수가 필요해져버린다. 따라서, `toString()`은 유연성을 위해 간단한 문자열 표현으로만 사용해야 한다.

결국 제 코드에서는 `toString()`을 가볍게 사용할만한 이유가 없어서, 전부 지우고 `OutputView`로 책임을 옮기게 되었습니다. 

제가 조사를 통해 이해한 내용이 맞는지 궁금합니다.

pci2676 replied:
굉장히 꼼꼼히 조사해보셨군요!  
크 쌍따봉드립니다 👍 👍 

view의 요구사항 변화가 domain 코드의 직접적인 변화를 이끌어내지 않는것이 저도 좋다고 생각합니다.  
이해하신 내용이 저의 생각과 일치하는것 같습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,69 @@
+    public Cars(NumericGenerator generator, String input) {
+        this.generator = generator;
+
+        isValidInput(input);
```
is~ 와 같은 네이밍은 맞다/아니다 와 같은 응답을 할 것 같은 네이밍으로 보입니다.

메서드의 책임과 더 관련있는 네이밍을 고려해보시면 좋을것 같습니다.

BurningFalls replied:
메서드가 "올바른 입력인지 확인"하는 역할을 수행하므로, `isValidInput`에서 `checkIfValidInput`으로 바꾸어 보았습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,69 @@
+                .collect(Collectors.toList());
+    }
+
+    public void isValidInput(String input) {
+        String regex = "^[^,]+(,[^,]+)*$";
+
+        Pattern pattern = Pattern.compile(regex);
+        Matcher matcher = pattern.matcher(input);
+
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("[ERROR] 문자열의 형태가 유효하지 않습니다.");
+        }
+    }
+
+    public List<String> parseWithComma(String input) {
+        return Arrays.stream(input.split(","))
+                .map(String::trim)
+                .filter(str -> !str.isEmpty())
+                .collect(Collectors.toList());
+    }
```
Cars의 역할과 책임에 맞는 구현인지 한번 고민해보시면 좋을것 같습니다.

BurningFalls replied:
`Cars`는 `List<Car>`을 다루는 클래스이기 때문에, 유저 입력을 처리하는 메서드를 포함하는 것은 어울리지 않는다는 점을 알게 되었습니다. 따라서 입력을 다루는 클래스인 `InputView`에 해당 메서드들을 옮겼습니다.

pci2676 commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,36 @@
+        @Test
+        @DisplayName("이름이 5자 초과 시 예외 발생")
+        void testNameUnderFive() {
+            String given = "문자열이일곱자";
```
6글자일때 문제가 발생하지 않을것이라는 것을 보장할 수 있는 테스트일까요?

BurningFalls replied:
이름이 6자 이상일 경우 예외가 발생하는지를 테스트하는 메서드이므로, 확실하게 경계값을 테스트해봐야 한다는 점 알게 되었고, 테스트 목적에 맞게 바꾸어보았습니다.

pci2676 commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,36 @@
+        @Test
+        @DisplayName("차 1칸 이동")
+        void testMove() {
+            Car car = new Car("차");
+            car.move();
+            int actual = car.getPosition();
+            int expected = 1;
+            assertThat(actual).isEqualTo(expected);
```
공백도 코드를 읽는데 도움을 주는 요소입니다.

공백을 활용해서 코드의 맥락을 파악하기 쉽게 분리해보시는것도 좋을것 같습니다.

BurningFalls replied:
* 변수 선언 초기화 + 메서드 사용
* `actual`과 `expected`를 선언
* `assert` 사용

총 세 부분으로 나누어 전체 테스트 코드를 공백으로 분리해서, 맥락을 쉽게 파악할 수 있도록 리팩터링해보았습니다.

pci2676 commented on src/test/java/racingcar/TryCountTest.java:
```diff
@@ -0,0 +1,56 @@
+    @Nested
+    class Feature {
+        @Test
+        @DisplayName("trycount 소비")
```
다른 테스트에 잘 표현된 이름이 많은것 같습니다.

아래 테스트도  
테스트 이름을 통해 테스트의 의미를 파악할 수 있도록 조금 더 풍부하게 표현해주시면 좋을것 같습니다.

BurningFalls replied:
코드의 리팩토링만 신경쓰고, 한글 설명에는 상대적으로 덜 신경썼던 것 같습니다. 앞으로는 테스트의 의미를 제대로 파악할 수 있게 `DisplayName`에도 많은 신경을 쓰겠습니다.

pci2676 commented on src/test/java/racingcar/CarsTest.java:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        @Test
+        @DisplayName("폭포가 4 포케가 4, 무빈이 3이 나오면 우승자는 폭포와 포케여야 함")
```
폭포, 포케, 무빈 등의 이름은 테스트 코드에 사용된 특정 값을 가리키고 있는것 같습니다.

코드가 변경됐을때 위 테스트 이름도 동일하게 변경할 수 있도록 강제할 수 없을것 같습니다.

테스트 설명을 조금 더 추상적으로 표현해보시면 좋을것 같습니다.

BurningFalls replied:
설명에 특정 값을 가리키는 문자열을 넣으면, 테스트 코드의 유연성을 해친다는 사실을 인지했습니다. 테스트코드에서 사용하는 메서드의 역할과 책임에 맞게 설명을 바꾸어보았습니다.

pci2676 commented on src/main/java/racingcar/controller/NumericGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar.controller;
+
+public interface NumericGenerator {
```
👍

BurningFalls replied:
함수형 인터페이스를 사용하여 이를 `implement`하는 여러 `class`를 만드는 종합적인 과정은 페어분이 도와주셨습니다. 저는 아직 이에 대한 지식이 많이 부족하여 열심히 공부해보려고 합니다.

pci2676 commented on src/main/java/racingcar/controller/numericgenerator/RandomNumericGenerator.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Random;
+
+public class RandomNumericGenerator implements NumericGenerator {
```
👍

BurningFalls replied:
위와 마찬가지로, 해당 개념에 대해서 공부해보려 합니다. 다만 아직은 이 부분의 코드를 유연하게 사용할 수 없어서, 혼자 리팩터링할 때는 다른 방법을 통해 랜덤값을 생성해보면서, 두 방법 모두 연습할 계획입니다.

pci2676 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private void validate(String name) {
+        nameUnderFive(name);
```
`nameUnderFive`라는 이름으로 검증을 수행한다는 느낌이 잘 들지 않는것 같습니다.

BurningFalls replied:
`validate`를 지우고 `checkIfNameUnderFive`로 어떤 동작을 수행하는지 확실하게 메서드 이름에 표기하였습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Override
+    public String toString() {
```
`Car`에서의 리뷰와 동일합니다.

BurningFalls replied:
`Car`과 함께 바꾸었습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,69 @@
+        return cars.stream()
+                .filter(car -> car.getPosition() == maxPosition)
+                .map(Car::getName)
+                .collect(Collectors.joining(", "));
```
`Car`에서의 리뷰와 동일합니다.

pci2676 commented on src/main/java/racingcar/model/TryCount.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    private void isOverZero(int number) {
```
앞선 네이밍 리뷰와 동일합니다.

BurningFalls replied:
`checkIfOverZero`로 변경하여 메서드 사용부만 봐도 메서드의 동작을 알 수 있게 표기했습니다.

pci2676 commented on src/main/java/racingcar/model/TryCount.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private void isOverZero(int number) {
+        if(number < 0) {
```
0이라는 숫자가 TryCount에서 의미있게 사용되는 것 같습니다.

앞선 리뷰를 참고해주세요.

BurningFalls replied:
0이라는 매직 넘버를 `END_LINE`이라는 상수로 만들어 주었습니다. 또한, 이를 사용하는 메서드 이름들도 그에 맞게 개선했습니다.

pci2676 commented on src/test/java/racingcar/CarsTest.java:
```diff
@@ -0,0 +1,97 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class CarsTest {
```
CarsTest로 많은 부분을 꼼꼼하게 테스트해주신것 같습니다. 👍 

다만 Cars에 Nested를 통해 서로 다른 성격의 테스트가 많이 있다는건 그만큼 Cars가 많은 책임을 가지고 있다는 반증인것 같습니다.

책임을 가볍게 만들어 보시면 좋을것 같습니다.

BurningFalls replied:
`Cars` 클래스에 있던 `checkIfValidInput`과 `parseWithComma` 메서드를 `InputView` 클래스로 옮김으로써, `Cars`의 책임을 가볍게 만들 수 있었습니다.

pci2676 commented on docs/README.md:
```diff
@@ -0,0 +1,21 @@
+
+- [X] 경주할 자동차 이름 입력받기: InputView.inputRacingCars
+- [X] 자동차 이름 목록 콤마를 기준으로 분리: Cars.parseWithComma
+- [X] 올바른 형식의 문자열인지 확인: Cars.validate
```
예를들어 입력의 형태는 다양한 방식으로 이루어질 수 있을것 같습니다.

Cars말고 입력을 다루는 곳에서 해결해보면 좋을것 같습니다.

BurningFalls replied:
앞서 남겨주는 리뷰에서 아이디어를 얻어, 두 메서드를 `InputView`로 옮길 수 있었습니다.

pci2676 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,39 @@
+package racingcar.model;
+
+public class Car {
+    private String name;
```
name은 값이 변하는 경우가 있을까요?

BurningFalls replied:
초기화를 제외하고 변하지 않는 객체에 대해서 확실하게 `final`을 선언해야 하는데, 아직 제대로 습관을 들이지 못한 것 같습니다. 변수 구상 단계에서 해당 변수가 변하지 않아야 하는지에 대해서 생각해보는 습관을 길러보도록 하겠습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.stream.Collectors;
+
+public class Cars {
+    private NumericGenerator generator;
+    private List<Car> cars;
```
이 멤버변수들도 값이 교체되는 경우가 있을까요?

BurningFalls replied:
앞에서와 마찬가지로 `final`을 확실하게 선언해주었습니다. 변수 구상 단계에서, 값이 교체되는지의 여부를 한번씩 생각해보고 넘어가도록 하겠습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public void moveCars() {
+        for (Car car : cars) {
+            if (generator.generate() >= 4) {
+                car.move();
+            }
+        }
```
들여쓰기가 2번 들어간것 같아보입니다.

조금 더 간결하게 표현할 수 없을까요?

그리고 4는 의미를 가지는 숫자 같아보입니다.

BurningFalls replied:
아직 `stream API`의 사용이 익숙하지 않다보니,  해당 코드에 대해서 적합한 방법을 찾지 못해 반복문으로 남겨놓게 되었습니다. 리팩토링하는 과정에서 검색을 통해 정답을 찾고 코드를 교체했어야 했는데, 못보고 지나가버렸습니다. 리뷰를 받고 제대로 검색해서 답을 찾을 수 있었습니다.

4라는 매직 넘버를 `MOVE_LINE` 상수로 바꾸었습니다.

pci2676 reviewed (changes requested):
반갑습니다. 🤠 

전체적으로 간결하고 깔끔하게 작성해주신것 같습니다.

조금 더 고민해볼만한 부분과 개선해볼만한 부분에 리뷰를 남겨두었으니 확인해주시고 다시 요청 부탁드리겠습니다.

BurningFalls committed:
style: CarController의 인스턴스 변수를 지역 변수로 변경

해당 인스턴스 변수들은 startGame() 메서드 안에서만 쓰이기 때문에, 인스턴스 변수가 아닌 startGame() 메서드 내부의 지역 변수로 선언해도 충분함.

BurningFalls committed:
style: 필요없는 공백 삭제

BurningFalls committed:
style: 매직넘버를 상수로 변경하고 Constants class로 이동

5라는 매직넘버는 코드를 보는 사람 입장에서 무슨 의미인지 해석하는 시간이 필요하기 때문에, 명시적으로 상수로 표현하는 것이 좋음(MAX_CAR_NAME_LENGTH).
상수로 표현함에 따라 상수를 모아서 따로 관리하는 Constants class를 만들고 그 class로 이동시킴

BurningFalls committed:
style: 메서드명을 메서드의 책임에 맞게 변경

isValidInput을 checkIfValidInput으로 변경

BurningFalls committed:
style: Cars 클래스의 역할과 책임에 맞지 않는 메서드를 Inputview 클래스로 이동

checkIfValidInput, parseWithComma 메서드를 이동

BurningFalls committed:
style: 메서드의 이동에 따른 Cars 생성자 초기화 방법 변경

기존에는 String을 받아 Cars 생성자에서 메서드로 처리했으나, InputView에서 한번에 처리하기 때문에 받는 방식을 변경

BurningFalls committed:
style: 생성자 초기화 방법 변화에 따라, 생성자 사용 부분 코드 수정

기존에 인수 타입이 String에서 List<String>으로 변경되었으므로, 전달하는 값도 이에 맞게 변경

BurningFalls committed:
style: 올바른 입력인지 확인하는 코드를 InputViewTest로 이동

BurningFalls committed:
test: 콤마를 기준으로 문자열 분리하는 테스트를 작성

BurningFalls committed:
style: 바뀐 생성자 초기화 방법에 맞게 인수 형식 수정

BurningFalls committed:
style: 테스트 메서드의 의도에 맞게 6글자 문자열을 테스트

BurningFalls committed:
style: 테스트 코드를 공백 활용해서 확실하게 분리

BurningFalls committed:
style: 테스트 메서드의 이름을 명확하게 표기

BurningFalls committed:
style: 테스트 메서드 설명의 유연성 확장

BurningFalls committed:
style: 메서드의 이름 명확하게 표기

BurningFalls committed:
style: 메서드의 이름 명확하게 표기

BurningFalls committed:
style: 게임이 끝나는 지점 (trycount가 0이 되는 지점)을 `END_LINE`으로 정의

BurningFalls committed:
docs: 개선한 사항(메서드 이름 변경)들을 README에 반영

BurningFalls committed:
style: 불변 객체에 final 선언

BurningFalls committed:
style: 반복문을 stream API로 교체

BurningFalls committed:
style: 매직 넘버(4)를 상수(MOVE_LINE)으로 정의

BurningFalls committed:
style: toString 삭제 get 사용 -> domain의 view 의존성 삭제

BurningFalls committed:
style: 불변 객체 final로 지정

BurningFalls committed:
style: 오타 수정

BurningFalls committed:
style: toString 삭제에 따른 테스트 메서드 변경 (get 사용)

BurningFalls committed:
style: printWinner 메서드 view 의존성 제거

BurningFalls committed:
style: calculateWinner 메서드 변경에 따른 테스트 코드 변경

BurningFalls commented:
안녕하세요, 먼저 빠른 리뷰 감사합니다!

세세하게 집어주신 덕분에, 제가 미처 생각하지 못하고 놓쳤던 부분들에 대해서 깊게 고민해보는 시간을 가질 수 있었습니다.

앞으로 코딩하면서, 아래와 같은 내용들을 계속 생각해보고자 노력하겠습니다.

* 매직 넘버 대신 상수 사용하기
* 변수의 불변성을 체크하고 `final` 키워드 사용하기
* 테스트 메서드의 `DisplayName` 정확하게 작성하기
* 인스턴스 변수를 (가능하다면) 지역변수로 만들기
* 클래스의 역할과 책임에 따라 메서드 분리하기

Reply에 적었던 내용처럼 `toString()`에서의 domain의 view 의존성에 대해서 제가 공부해서 이해한 내용이 맞는지 확인해주셨으면 좋겠습니다. 

또한, 이번이 제가 처음으로 해보는 코드리뷰라서, 혹시 코드리뷰 과정 자체에 대해서 제가 잘못하고 있는 부분, 개선했으면 하는 부분이 있다면, 알려주시면 발전의 밑거름으로 삼을 수 있을 것 같습니다.

다시 한번 세세한 코드리뷰 감사드립니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.stream.Collectors;
+
+public class Cars {
+    private final NumericGenerator generator;
```
NumericGenerator을 Cars가 제공하는 모든 메서드에서 참조하고 있을까요?

그렇지 않다면 응집도가 높은 코드라고 볼 수 있을지 고민해보시면 좋을것 같습니다.

BurningFalls replied:
추상화를 적용시키지 않은 채로 코드를 바꾼다면 `moveCars`에서 인수로 `generator`를 받도록 바꿀 수 있을 것 같습니다. 

그러나, 밑에서 조언해주신대로 `MoveStrategy` 인터페이스를 만든다면, `Cars` 클래스에서 `generator` 변수를 완벽하게 분리시키면서 `moveCars`에는 인수로 `MoveStrategy`를 받는 형식으로 깔끔하게 바꿀 수 있을 것 같습니다.

그래서 이 부분 또한 나중에 추상화를 시키면서 한번에 바꿔보고자 합니다.

pci2676 commented on src/main/java/racingcar/constants/Constants.java:
```diff
@@ -0,0 +1,7 @@
+package racingcar.constants;
+
+public class Constants {
+    public static final int MAX_CAR_NAME_LENGTH = 5;
+    public static final int END_LINE = 0;
+    public static final int MOVE_LINE = 4;
+}
```
상수는 사용되는 곳의 맥락과 함께 선언되어있는 것이 이해를 도와 좋을것 같습니다.

상수만을 위한 상수 클래스는 제거해주세요.

BurningFalls replied:
상수를 사용하는 클래스에 상수가 정의되어 있는 것이 가시성이 훨씬 좋네요!

다만, 한 가지 의문점이 생겼습니다.
만일 동일한 상수를 여러 클래스에서 사용하는 경우, 각각의 클래스에 동일한 상수 정의 코드가 생기는데, 이를 허용하는게 맞는 것인지 아니면 해결할 수 있는 단순한 방법이 있는 것인지 궁금합니다.

gpt의 답변은 공통 인터페이스 정의, 유틸리티 클래스 사용, 상속의 세 가지가 있었는데 오버엔지니어링이라는 생각이 듭니다. 게다가 기존에 함께 선언해서 이해를 돕는다는 목적에 어긋나게 되는 것 같습니다.

제가 생각해본 방법은 "애초에 하나의 상수를 여러 곳에서 사용하는 일이 없도록 클래스들을 잘 구성한다" 정도인 것 같습니다.

pci2676 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,47 @@
+
+    public void moveCars() {
+        cars.stream()
+                .filter(car -> generator.generate() >= Constants.MOVE_LINE)
```
`generator.generate() >= Constants.MOVE_LINE`의 구현도 나쁘지 않은데요.

지금은 움직임을 판단하는 특정 방법으로 숫자값을 판단하고 있지만  
움직일수 있는가? 없는가? 라는 관점으로 보았을때 조금 더 추상화해볼수 있지 않을까. 하는 고민이 되는것 같습니다.

꼭 반영을 하실 필요는 없지만 한번쯤은 고민해보셔도 좋을것 같습니다.

BurningFalls replied:
```java
public interface MoveStrategy {
   boolean canMove();
}
```

```java
public class RandomMoveStrategy implements MoveStrategy {
   // ...
   @Override
   public boolean canMove() {
      return generator.generate() >= Constants.MOVE_LINE;
   }
}
```

```java
public void moveCars(MoveStrategy moveStrategy) {
   cars.stream()
      .filter(car -> moveStrategy.canMove())
      .forEach(Car::move)
}
```

이렇게 추상화하는게 맞나요? 아직 `interface` 사용이 익숙하지 않아서 gpt의 도움을 받아보았습니다. 이렇게 하면 다음과 같은 장점이 있을 것 같습니다.

* 움직임을 판단하는 방법이 '숫자값'이 아닌 다른 기준이 생기면(ex. 시동이 켜져있는지), `EngineMoveStrategy` 클래스를 만들어서 `canMove()`에 새로운 기준을 구현하여 적용시킬 수 있다.

다만, 아직 익숙하지 않은 개념이라 수정하는 경우 정해진 시간까지 `step2`를 마무리하기 힘들 것 같아, 기한 내 주어진 과제를 마무리하고 이후에 적용시켜보고자 합니다.

pci2676 commented on src/main/java/racingcar/model/TryCount.java:
```diff
@@ -0,0 +1,39 @@
+
+import racingcar.constants.Constants;
+
+public class TryCount {
```
TryCount는 일종의 값 객체로써 표현하신것 같습니다.

보통의 경우 값 객체는 불변하게 설계하는 경우가 많으니 이 점도 한번 쯤 고민해보시면 좋을것 같습니다.

BurningFalls replied:
값 객체를 `final` 클래스로 선언하는 것에 대해 새롭게 알게 되었습니다.
다만, 제 코드에서는 `TryCount`의  `value`를 `consume()` 메서드를 통해 소비하고 있기 때문에 `final` 선언이 불가능했습니다.

따라서, 게임을 `tryCount.value`번 실행할 때 `consume()`을 사용하는 것이 아니라, `tryCount.value`번 `index`를 증가시키는 반복문으로 바꾸어 객체의 불변성을 충족시켰습니다.

이 리뷰를 통해 다음과 같은 문제를 생각해보았습니다.
> 값 객체를 다루는 클래스를 반드시 불변으로 설계해야 하는가?

이에 대해서는 아래와 같이 정리했습니다.
* 불변성은 스레드 안전성, 일관성 있는 상태 유지, 재사용성 면에서 굉장히 큰 장점을 가진다.
* 다만 그 값이 매우 자주 변한다면, 값 변경마다 새로운 객체를 생성하는 것은 메모리와 코드의 복잡성에 있어 비효율적이다.

따라서 제 코드에서는 단순히 반복문으로 바꾸기만 하면 불변 객체의 장점을 취할 수 있기 때문에, 불변 객체로 바꾸었습니다.

pci2676 reviewed (approved):
리뷰에 답변 남겨주신 부분에 코멘트를 남겨두었고 추가로 조금 리뷰를 남겨두었습니다.

다음 단계를 진행하면서 같이 고려해주시면 좋을것 같습니다.

꼼꼼한 정리까지 좋았습니다 👍 

머지하겠습니다. 🚀

pci2676 commented:
>매직 넘버 대신 상수 사용하기
변수의 불변성을 체크하고 final 키워드 사용하기
테스트 메서드의 DisplayName 정확하게 작성하기
인스턴스 변수를 (가능하다면) 지역변수로 만들기
클래스의 역할과 책임에 따라 메서드 분리하기

인스턴스 변수를 지역변수로 변경하는것은 클래스내에서 해당 인스턴스 변수가 얼마나 응집도 있게 사용되는지에 대해 고민해보시면 좋을것 같습니다 😄

BurningFalls commented:
말씀해주신 부분들에 대해서 조사하고 분석한 결과를 작성해보았습니다. 소중한 피드백 부탁드립니다. 감사합니다.

> 인스턴스 변수를 지역변수로 변경하는것은 클래스내에서 해당 인스턴스 변수가 얼마나 응집도 있게 사용되는지에 대해 고민해보시면 좋을것 같습니다 😄

인스턴스 변수로 만들지, 지역변수로 만들지에 대한 기준이 "해당 변수가 얼마나 응집도 있게 사용되는지"임을 확인했습니다. 앞으로 클래스의 책임에 있어 변수가 어떤 역할을 맡고 있고 얼마나 연관성있는지를 지속적으로 확인해보도록 하겠습니다.
